pennylane/data/dataset.py

```diff
@@ -151,5 +149,54 @@ def from_dataset(cls, dataset):
             >>> print(vars(new_dataset))
             {'dtype': None, '__doc__': '', 'kw1': 1, 'kw2': '2', 'kw3': [3]}
         """
-        kwargs = {key: val for (key, val) in dataset.__dict__.items() if key not in ["dtype"]}
-        return cls(dtype=dataset.dtype, **kwargs)
+        return cls(**dataset.attrs)
+
+
+class RemoteDataset(Dataset):
+    """A dataset object that provides additional utility, given that the data comes
+    from the PennyLane remote dataset source."""
+
+    def __init__(self, dtype, folder, attr_prefix, **kwargs):
+        self._dtype = dtype
+        self._folder = folder.rstrip("/")
+        self._prefix = os.path.join(self._folder, attr_prefix.rstrip("/")) + "_{}.dat"
+        self.__doc__ = ""
+
+        self._fullfile = self._prefix.format("full")
+        if not os.path.exists(self._fullfile):
+            self._fullfile = None
+
+        for f in glob(self._folder + "/*.dat"):
+            print(f)
+            data = self._read_file(f)
+            for attr, value in data.items():
+                setattr(self, f"{attr}", value)
+
+        super().__init__(**kwargs)
+
+    def __get_filename_for_attribute(self, attribute):
+        return self._fullfile if self._fullfile else self._prefix.format(attribute)
+
+    def __getattribute__(self, name):
+        try:
+            return super().__getattribute__(name)
+        except AttributeError:
+            pass
+        filepath = self.__get_filename_for_attribute(name)
+        if os.path.exists(filepath):
+            # TODO: setattr?
+            return self._read_file(filepath)[name]
+        # TODO: download the file here?
+        raise AttributeError(
+            f"'Dataset' object has no attribute {name} and no matching file was found"
+        )
+
+    def setdocstr(self, docstr, args=None, argtypes=None, argsdocs=None):
```

Comment:
Not sure I understand this function.

Reply:
the docstring (when you run `help(my_data)`) will show docs based on the data actually available from the Dataset object. Because it's set dynamically, so is the docstring, so this function does that. As an example, if you download a "full" qchem dataset, you'll see a docstring like this:

<details><summary>Result from `help(my_data)`</summary>

```
Help on RemoteDataset in module pennylane.data.dataset:

<pennylane.data.dataset.RemoteDataset object>
    Quantum chemistry dataset.
    
    Args:
            molecule (<class 'pennylane.qchem.molecule.Molecule'>): Molecule object describing the chemical system
            hamiltonian (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Hamiltonian of the system using Jordan-Wigner mapping
            ham_wire_map (<class 'dict'>): Hamiltonian wire map post tapering
            sparse_hamiltonian (<class 'scipy.sparse._csr.csr_matrix'>): Sparse Hamiltonian of the system
            meas_groupings (<class 'tuple'>): Qubit-wise commuting measurement groupings
            symmetries (<class 'list'>): Symmetries of the hamiltonian used to taper qubits
            paulix_ops (<class 'list'>): Pauli-X operators used in buiding Clifford unitary for tapering
            optimal_sector (<class 'list'>): Eigensector containing ground state of the tapered Hamiltonian
            num_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Number operator of the system
            dipole_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Dipole moment operator of the system
            spin2_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Total spin operator of the system
            spinz_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Spin projection operator of the system
            hf_state (<class 'pennylane.numpy.tensor.tensor'>): Hartree-Fock state for the system
            fci_energy (<class 'numpy.float64'>): FCI ground state energy
            tapered_hamiltonian (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Tapered hamiltonian operator
            tapered_dipole_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Tapered dipole moment operator
            tapered_num_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Tapered number operator
            tapered_spin2_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Tapered total spin operator
            tapered_spinz_op (<class 'pennylane.ops.qubit.hamiltonian.Hamiltonian'>): Tapered spin projection operator
            tapered_hf_state (<class 'pennylane.numpy.tensor.tensor'>): Tapered Hartree-Fock state
            vqe_energy (<class 'pennylane.numpy.tensor.tensor'>): Variational energy given by the vqe_circuit for provided vqe_params
            vqe_params (<class 'pennylane.numpy.tensor.tensor'>): Variational parameters for which ansatz evolve to ground state
            vqe_circuit (<class 'functools.partial'>): Variational circuit for obtaining ground-state
    
    Returns:
            Dataset(qchem)
```
</details>

Though I think there's work to be done here. Users will be able to dynamically add new data to an existing dataset, but it won't add to the docstring. @obliviateandsurrender what if instead of all this custom logic, we _always_ set the docstring to include all attributes, but find some other way to indicate that it might not be available? In fact, it is available if we dynamically fetch files from s3 using the __getattribute__ overload! As I type this, I think that's my preferred behaviour.

Reply:
This feature seems to me as a nice-to-have feature. Let's not spend too much time on it, at this moment.